### PR TITLE
chore(repo): Major tech debt

### PR DESCRIPTION
**Pull request by itsdevbear**
Improvements:
- [x] introduce `pond` workers, move jobs to operate on the a `producer consumer` software design pattern. 
- [x] introduce `metrics`
- [x] Decouple `job` package from `worker` package, generic should help. Then pass in type at instantiation. (in baseapp / jm) 
- [x] Standardize job interfaces. 
- [x] ~~Move `JobManager` out of the `baseapp`? Maybe?~~ Leaving it here.
- [x] Handle all jobs using new producer consumer pattern. 
- [x] Improve `cmd` package, handle config management better. It's too hard to use rn, very opinionated. 
- [x] ~~Introduce `core` package? standardize interfaces. ~~
- [x] Standardize Template App 
- [x] Panic Handler for Job
- [x] Handle context cancellations correctly.

Handle in future PR
- [ ] Refactor subscription based jobs to abstract away `chan` usage and decouple from `go-ethereum` 
- [ ] Build General Retry Object (that can take a function and retry it)
- [ ] Document Architecture

### CODE REVIEW

File: worker/metrics.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package worker
+
+import "github.com/prometheus/client_golang/prometheus"
+
+func (p *Pool) setupMetrics(prefix string) {
```
Comment (transmissions12) [inline worker/metrics.go:5]: bullish

File: cmd/start.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+			// Build the application, then start it.
+			app.Setup(ab, log.NewBlankLogger(cmd.OutOrStdout()))
+			if err = app.Start(ctx); err != nil {
+				return err
+			}
+
+			// Wait for the context to be done.
+			<-ctx.Done()
+			// TODO: should we return error here based on ctx.Err()?
```
Comment (calbera) [inline cmd/start.go:59]: yeah just return ctx.Err()

File: worker/pool.go [hunk 3]
```diff
@@ -1,107 +1,29 @@
...
-		)
-		p.workers = append(p.workers, w)
+// and other functionality to the pool.
+type Pool struct {
+	name string
+	*pond.WorkerPool
+}
+
+// NewPool creates a new pool.
+func NewPool(ctx context.Context, cfg *PoolConfig) *Pool {
```
Comment (calbera) [inline worker/pool.go:16]: doesn't need context

File: baseapp/manager.go [hunk 4]
```diff
@@ -2,113 +2,164 @@ package baseapp
...
+func (jm *Manager) Logger(ctx context.Context) log.Logger {
+	return sdk.UnwrapSdkContext(ctx).Logger().With("namespace", "job-manager")
+}
+
+// Start calls `Setup` on the jobs in the registry as well as spins up
+// the worker pools.
+func (jm *Manager) Start(ctx context.Context) {
+	// We pass in the context in order to handle cancelling the workers.
+	jm.jobExecutors = worker.NewPool(ctx, jm.executorCfg)
+	jm.jobProducers = worker.NewPool(ctx, jm.producerCfg)
```
Comment (calbera) [inline baseapp/manager.go:81]: dont need context here

File: baseapp/manager.go [hunk 5]
```diff
@@ -2,113 +2,164 @@ package baseapp
...
+						jm.Logger(ctx).Error("error in subscription", "err", err)
 						// TODO: add retry mechanism
 						ethSubJob.Unsubscribe(ctx)
 						return
 					case val := <-ch:
-						jm.executionPool.AddTask(job.NewExecutor(ctx, ethSubJob, val))
+						jm.jobExecutors.Submit(workertypes.NewPayload(ctx, ethSubJob, val).Execute)
 						continue
 					}
 				}
```
Comment (calbera) [inline baseapp/manager.go:159]: calling to `Subscribe` and `EthSubscribe` should have diff underlying contexts.

File: baseapp/baseapp.go [hunk 6]
```diff
@@ -56,23 +59,40 @@ func (b *BaseApp) Logger() log.Logger {
...
-func (b *BaseApp) Start() {
+func (b *BaseApp) Start(ctx context.Context) error {
 	b.Logger().Info("starting app")
 
-	// TODO: create a new context for every job request / creation.
-	ctx := sdk.NewContext(
-		context.Background(),
+	// Wrap the context in sdk.Context in order to attach our clients, logger and db.
+	// TODO: is this bad practice we are just stealing from the cosmos sdk?
+	ctx = sdk.NewContext(
```
Comment (calbera) [inline baseapp/baseapp.go:67]: we need to really make sure that every call to a job uses a diff underlying context for all .Setup(), .Subscribe()/.EthSubscribe(), and .NewPayload()

File: worker/pool.go (hunk 3 above)
Comment (itsdevbear) [reply worker/pool.go:16]: have to pass the context down to the pool.

File: baseapp/manager.go (hunk 4 above)
Comment (itsdevbear) [reply baseapp/manager.go:81]: have to pass to the underlying pool (for now until we handle new contexts and stuff)

File: baseapp/baseapp.go (hunk 6 above)
Comment (itsdevbear) [reply baseapp/baseapp.go:67]: agree, just haven't gotten there yet.

**Comment by itsdevbear** [comment]
@calbera do we wanna merge this and then just continue to cleanup / iterate? Or keep it open until all is good?

**Comment by calbera** [comment]
you can merge idrc, just we shouldn't update bts until this is fully good

**Comment by itsdevbear** [comment]
it's stable, just not all things cleaned up. external APIs won't change much. But whatever you guys think is best for `bts`

File: cmd/start.go (hunk 2 above)
Comment (itsdevbear) [reply cmd/start.go:65]: nah doesnt work, added an if

**Comment by itsdevbear** [comment]
@ocnc @calbera @transmissions12

File: baseapp/manager.go (hunk 5 above)
Comment (itsdevbear) [reply baseapp/manager.go]: going to handle in a later PR tbh.

**Pull request merged**